added mainnet fork arbitrum test

crisgarner opened the pull request:
test with:

``` forge test --match-path ./test/arbitrum/JPEGzMainnetOracle.t.sol --fork-url INFURA_URL```

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+  ChainlinkOracle ethOracle =
+    new ChainlinkOracle(address(ethAggregator), address(this));
+  address weth = 0x82aF49447D8a07e3bd95BD0d56f35241523fBab1;
+  address user = address(0x1);
```
Please change this address as this is a precompile address. 
I use something like address(0x51);

voith replied:
Please change this address as this is a precompile address. 
I use something like `address(0x51)`.

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+  uint256 burnFee = 150;
+  uint256 mintFee = 150;
+  uint256 liquidationPenalty = 20;
+  address treasury = address(0x3);
```
same as above

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+  function testConstructor_ShouldRevert_WhenBurnFeeIsHigh(uint256 _burnFee)
+    public
+  {
+    if (_burnFee > 1000) {
```
Forge has a trick for fuzzing: `vm.assume(_burnFee > 1000)`.
With this trick, you should be able to get rid of the if block.

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+    assertEq(address(ethOracle), address(ethVault.collateralPriceOracle()));
+    assertEq(address(ethOracle), address(ethVault.ETHPriceOracle()));
+    assertEq(treasury, ethVault.treasury());
+    assertEq(0 ether, ethVault.minimumTCAP());
```
wish we had refactored stuff like this which says TCAP :(

crisgarner replied:
I agree, especially when we try to make tests it throws an error that JPGEZ != TCAP. so in the end we end using TCAP contract

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+    );
+
+    if (!(_burnFee > 1000)) {
+      assertEq(_burnFee, vault.burnFee());
```
This should be a separate test. The test name is `ShouldRevert_WhenBurnFeeIsHigh`. 
Please split these into two separate tests.

crisgarner replied:
Fixed, this is already tested on the testConstructor_ShouldSetParams_WhenInitialized

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+  function testConstructor_ShouldRevert_WhenMintFeeIsHigh(uint256 _mintFee)
+    public
+  {
+    vm.assume(_mintFee > 1000);
```
seems like you already know about `vm.assume` :)

crisgarner replied:
yeah, seems like I had some old test and forgot to update them

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+    uint256 _liquidationPenalty,
+    uint256 _ratio
+  ) public {
+    if (_liquidationPenalty + 100 < _liquidationPenalty) {
```
Can we use `vm.assume(_liquidationPenalty + 100) >= _ratio)`?

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+    );
+
+    if (!((_liquidationPenalty + 100) >= _ratio)) {
+      assertEq(_liquidationPenalty, vault.liquidationPenalty());
```
Please split this as a separate as this test is different from its name i.e `ShouldRevert_WhenLiquidationPenaltyIsHigh`.

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+  function testSetLiquidationPenalty_ShouldUpdateValue(
+    uint256 _liquidationPenalty
+  ) public {
+    if (_liquidationPenalty + 100 < _liquidationPenalty) {
```
can we use `vm.assume` here?

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+  }
+
+  function testShouldUpdateRatio(uint256 _ratio) public {
+    if (_ratio < 100) {
```
can we do `vm.assume((_ratio < 100) &&(ethVault.liquidationPenalty() + 100 >= _ratio) )`?

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+    assertEq(ethVault.minimumTCAP(), _minimumTCAP);
+  }
+
+  function testMint_ShouldCreateTCAP_WhenFeeIsPaid() public {
```
```suggestion
  function testMint_ShouldCreateJPEGZ_WhenFeeIsPaid() public {
```

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+    orchestrator.setRatio(ethVault, _ratio);
+  }
+
+  function testSetMinimumTCAP_ShouldUpdateValue(uint256 _minimumTCAP) public {
```
```suggestion
  function testSetMinimumJPEGZ_ShouldUpdateValue(uint256 _minimumTCAP) public {
```

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+    assertEq(collateral, 10 ether);
+    assertEq(debt, 1 ether);
+    assertEq(owner, user);
+    assertEq(address(treasury).balance, fee);
```
Just to be sure that our calculations with fees are correct, can you please check that this actually 1.5% of 1 ether?
i.e 
```solidity
unit256 _fee = ((15/10) * 1 ether * JPEGZ_PRICE())/(100 * ETH_PRICE());
assertEq(address(treasury).balance, _fee);
```
using 15/10 as we know that is equal 1.5. I just want to make sure that we've set `uint256 mintFee = 150;` correctly above.

crisgarner replied:
added these lines:

```
uint256 _fee = ((15 / 10) * 1 ether * ethVault.getOraclePrice(jpegzOracle)) / (ethVault.oracleDigits() * 10 * ethVault.getOraclePrice(ethOracle));
assertEq(address(treasury).balance, _fee);

```

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+
+  function testBurnTCAP_ShouldBurn_WhenFeeIsPaid(uint96 _tcapAmount) public {
+    // checks
+    if (_tcapAmount < 1 ether) {
```
use `vm.assume`

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+    vm.startPrank(user);
+    uint256 requiredCollateral = ethVault.requiredCollateral(_tcapAmount);
+    uint256 mfee = ethVault.getBurnFee(_tcapAmount);
+    uint256 bfee = ethVault.getBurnFee(_tcapAmount);
```
Can you test the value of bfee like the way I've requested you to test the mintFee?

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+    public
+  {
+    //setUp
+    if (_newFee > 1000) {
```
use `vm.assume`

voith commented on test/arbitrum/JPEGzMainnetOracle.t.sol:
```diff
@@ -0,0 +1,505 @@
+    uint96 _amount
+  ) public {
+    //setUp
+    if (_mintFee > 1000) {
```
use `vm.assume`

voith reviewed (commented):
Please change the addresses that use precompile addresses. 
Also, I've requested some cleanup.

crisgarner commented:
Fixed the observations @voith. Fixed the test also. In order to test the fork now you need to set up the .env variable `FOUNDRY_PROFILE=fork` and disable it for normal testing.

fork is now tested with

`forge test --fork-url RPC`

voith reviewed (approved):
LGTM 👍
